### DISCUSSION TITLE
Different levels of fidelity of Floris

### DISCUSSION THREAD

**Discussion by FMenhorn** [NatLabRockies/floris#623] [Q&A] [answered]
Hi,

I am very new to Floris and want to thank you, first of all, for the great documentation of your code. It was quite easy to set up the code and get some first examples running.

I am working in the field of Uncertainty Quantification, specifically for Multilevel/Multifidelity methods and would like to test Floris (or Wind Energy in general) as an application. Is it possible in Floris to create different levels of fidelity with different computational cost? In other words, are there different levels of accuracy of the solution with are easily steerable, e.g., with the input file, which result in different computation times--- usually higher accuracy equals higher computational cost? In the area of multifidelity this is often achieved by changing the discretization of the grid or using different solvers with different accuracy. I played a bit around with different number of grid points/turbines/solver but could not find an obvious choice. 

Thanks a lot for your help!

**Comment by paulf81** [comment]
Hi, and thank you very much for the kind words about FLORIS.  I'm very glad to hear you had an easy time getting started!

So to your question, it's a nice one and probably off the top of my head I can only give an incomplete answer.  I think your first idea is also mine, changing the number of grid points per turbine.  Some of the models allow you to go all the down to 1 point per turbine, which does speed things up but should involve some variation in output.

Another place where there can be a clear trade in time/accuracy can be how many wind speeds and wind direction are included in an AEP calculation.  Start with say 0 - 25 m/s in 1 m/s steps and 0 - 360 in  1 deg steps, you can save a lot of time by going to larger steps and eventually things will start to vary.  On this topic you can find some literature looking at a good compromise between speed and precision independent of FLORIS specific.

I wanted to say your question reminds me of this paper:
https://iopscience.iop.org/article/10.1088/1742-6596/1037/4/042012/meta

Which might also give you some ideas of model parameter variations that could be considered

Tagging @jaredthomas68 in case has some other thoughts

**Reply by FMenhorn** [reply]
Thanks Paul for the quick reply. Indeed I also want to look into optimization, thanks for the paper.

I played a bit more around today. It looks like varying the turbine_grid_points between 1-3 points does not change the runtime a lot. I assume the difference in grid points is not large enough (since its only a 2D grid per turbine). Is it also possible to use FlowFieldGrid?
If I understand it correctly it creates a grid over the full 3D domain. I set the attribute flow_field_grid_points in the yaml-file, but I get an error " AttributeError: 'FlowFieldGrid' object has no attribute 'sorted_indices' " when I try to run it. Is this grid still supported?
Thanks a lot!
Friedrich

**Reply by paulf81** [reply]
Hi Friedrich, right now FlowFieldGrid is mostly used in visualization and not in normal calculations, which use only points on turbines to optimize for speed.

WRT turbine_grid_points, I think if you scale up both the number of turbines and ws/wd, you will start to see this value matter more and more substantially.

**Reply by rafmudaf** [reply]
> I played a bit more around today. It looks like varying the turbine_grid_points between 1-3 points does not change the runtime a lot.

@FMenhorn I'm actually surprised that you didn't see much of a difference in run time. Can you verify that your method for changing the number of points is actually reflected? It would be helpful if you copy / paste a simple version of your script for changing the number of points and measuring the run time.

**Reply by FMenhorn** [reply]
Thanks a lot for the help to all of you! 
Yes, I can give you the example. This is the very simple example code:
--------
        fi = FlorisInterface("inputs/gch.yaml")
        D = 126.
        layout_x = np.array([0, D*5, D*10, D*15, D*2.5, D*7.5, D*12.5, 0, D*5, D*10, D*15])
        layout_y = [0, 0, 0, 0, D*5, D*5, D*5, D*8, D*8, D*8, D*8]
        fi.reinitialize(layout_x=layout_x, layout_y=layout_y)

        # Sweep wind speeds but keep wind direction fixed
        ws_array = [8.0]
        fi.reinitialize(wind_speeds=ws_array)

        # Define a matrix of yaw angles to be all 0
        # Note that yaw angles is now specified as a matrix whose dimesions are
        # wd/ws/turbine
        num_wd = 1
        num_ws = len(ws_array)
        num_turbine = len(layout_x)
        yaw_angles = np.zeros((num_wd, num_ws, num_turbine))

        # Calculate
        start = time.time()
        fi.calculate_wake(yaw_angles=yaw_angles)
        end = time.time()
       print(end - start)
--------
I simply adapted it from one of the example files, created a 11 turbine case that I found in a paper, and also use the example config file gch.yaml. In gch.yaml file I just played around with the turbine_grid_points but couldn't see a difference. I intentionally look only at one wind direction and speed. Sorry if I misunderstood something.

**Reply by paulf81** [reply]
I modified this a little bit to take more time:

```
import numpy as np
import time

from floris.tools import FlorisInterface    

fi = FlorisInterface("inputs/gch.yaml")
D = 126.
layout_x = np.array([0, D*5, D*10, D*15, D*2.5, D*7.5, D*12.5, 0, D*5, D*10, D*15])
layout_y = [0, 0, 0, 0, D*5, D*5, D*5, D*8, D*8, D*8, D*8]
fi.reinitialize(layout_x=layout_x, layout_y=layout_y)

# Sweep wind speeds but keep wind direction fixed
ws_array = np.arange(1.0,25.0)
wd_array = np.arange(0.0,360.0,1.0)
fi.reinitialize(wind_speeds=ws_array, wind_directions=wd_array)

# Define a matrix of yaw angles to be all 0
# Note that yaw angles is now specified as a matrix whose dimesions are
# wd/ws/turbine
num_wd = 1
num_ws = len(ws_array)
num_turbine = len(layout_x)
yaw_angles = np.zeros((num_wd, num_ws, num_turbine))

# Calculate
start = time.time()
fi.calculate_wake(yaw_angles=yaw_angles)
end = time.time()
print(end - start)
```

Then I ran 3 times with turbine_grid_points: 3, then 3 times with 1:

```
(floris) pfleming-37929s:examples pfleming$ python test_time.py 
2.4145960807800293
(floris) pfleming-37929s:examples pfleming$ python test_time.py 
2.5329341888427734
(floris) pfleming-37929s:examples pfleming$ python test_time.py 
2.4149789810180664
(floris) pfleming-37929s:examples pfleming$ python test_time.py 
0.3276028633117676
(floris) pfleming-37929s:examples pfleming$ python test_time.py 
0.3157048225402832
(floris) pfleming-37929s:examples pfleming$ python test_time.py 
0.32483887672424316
```

It could be that with 1 ws/wd the computation is fast enough that other things in the computation come to dominate.

**Reply by FMenhorn** [reply]
Sorry for the belated reply due to Easter break.

Thanks Paul! This indeed increased the computation time. I just swapped it around and sweep over all wind directions for a single wind speed but the increase in computation time is similar. (I do the change because I want to apply uncertainty on the wind speed in the study)

However, unfortunately, I don't get a big difference with respect to the power production when using the difference grids, turbine_grid_points=1 or turbine_grid_points=3. I guess the single grid point is already good enough in the approximation quality and the improvement turbine_grid_points=3 provides is very small. Hence, I am unsure if I can use it in my study. Thank you very much for your help though!

**Reply by Bartdoekemeijer** [reply]
Hi @FMenhorn. Check out #499 for a convergence (or lack thereof) study of the number of turbine grid points.

**Reply by paulf81** [reply]
Thanks @Bartdoekemeijer  for a helpful link!

**Reply by FMenhorn** [reply, marked as answer]
Yes, thanks @Bartdoekemeijer and @paulf81 ! This indeed helped me. (I had to modify the code a bit to not raise a ValueError for grid points > 3) Together with the other suggestions I was able to create three levels of fidelity:

- coarse: 1 turbine grid points + calculate_no_wake for the computation of the turbine power
- medium: 1 turbine grid points + calculate_wake for the computation of the turbine power
- fine: 40 turbine grid points + calculate_wake for the computation of the turbine power

I get a nice increase in runtime and convergence of the solution.

Thanks all for your help!

**Comment by FMenhorn** [comment]
Now, I have an additional question. I want to control the yaw angles and then visualize them. I can set the yaw angles as parameter for calculate_no_wake/calculate_wake. For the visualization using calculate_horizontal_plane, I have to set the same yaw angles as parameter in that function call as well, correct? It doesn't automatically know the yaw angle, I believe. Is there also a way to create a gif like you have on your page on the top left of https://nrel.github.io/floris/intro.html ?

Thanks again!

**Reply by Bartdoekemeijer** [reply]
@FMenhorn could you open a separate discussion for this question? I think more people will find it useful and it'll be easier to find that way. Thanks!

**Reply by rafmudaf** [reply]
@FMenhorn I've found the script I used to create that gif. Please point me to the new discussion and I'll post it there. Thanks!
